boards/beagle/beagleconnect_freedom/board_antenna.c

```diff
@@ -1,116 +1,119 @@
+	for (i = 0; i < ARRAY_SIZE(ant_gpios); i++) {
+		gpio_pin_configure_dt(&ant_gpios[i], 0);
+	}
 	return 0;
 }
 
-SYS_INIT(board_antenna_init, POST_KERNEL, CONFIG_BOARD_ANTENNA_INIT_PRIO);
-
+/**
+ * Custom TI RFCC26XX callback for switching the on-board antenna mux on radio setup.
+ */
 void board_cc13xx_rf_callback(RF_Handle client, RF_GlobalEvent events, void *arg)
 {
-	bool    sub1GHz   = false;
+	bool sub1GHz = false;
 	uint8_t loDivider = 0;
+	int i;
 
-	/* Switch off all paths first. Needs to be done anyway in every sub-case below. */
-	GPIO_setOutputEnableDio(BOARD_RF_HIGH_PA, GPIO_OUTPUT_DISABLE);
-	GPIO_setOutputEnableDio(BOARD_RF_SUB1GHZ, GPIO_OUTPUT_DISABLE);
+	/* Clear all antenna switch GPIOs (for all cases). */
+	for (i = 0; i < ARRAY_SIZE(ant_gpios); i++) {
+		gpio_pin_configure_dt(&ant_gpios[i], 0);
+	}
 
 	if (events & RF_GlobalEventRadioSetup) {
 		/* Decode the current PA configuration. */
-		RF_TxPowerTable_PAType paType = (RF_TxPowerTable_PAType)
-			RF_getTxPower(client).paType;
+		RF_TxPowerTable_PAType paType =
+			(RF_TxPowerTable_PAType)RF_getTxPower(client).paType;
 		/* Decode the generic argument as a setup command. */
 		RF_RadioSetup *setupCommand = (RF_RadioSetup *)arg;
 
 		switch (setupCommand->common.commandNo) {
-		case (CMD_RADIO_SETUP):
-		case (CMD_BLE5_RADIO_SETUP):
+		case CMD_RADIO_SETUP:
+		case CMD_BLE5_RADIO_SETUP:
 			loDivider = RF_LODIVIDER_MASK & setupCommand->common.loDivider;
-			/* Sub-1GHz front-end. */
-			if (loDivider != 0)
-				sub1GHz = true;
 			break;
-		case (CMD_PROP_RADIO_DIV_SETUP):
+		case CMD_PROP_RADIO_DIV_SETUP:
 			loDivider = RF_LODIVIDER_MASK & setupCommand->prop_div.loDivider;
-			/* Sub-1GHz front-end. */
-			if (loDivider != 0)
-				sub1GHz = true;
 			break;
 		default:
 			break;
 		}
+		sub1GHz = (loDivider != 0);
 
-		/* Sub-1 GHz */
-		if (paType == RF_TxPowerTable_HighPA) {
-			/* PA enable --> HIGH PA */
-			/* LNA enable --> Sub-1 GHz */
-			/* Note: RFC_GPO3 is a work-around because the RFC_GPO1 */
-			/* is sometimes not de-asserted on CC1352 Rev A. */
-			IOCPortConfigureSet(BOARD_RF_HIGH_PA,
-					IOC_PORT_RFC_GPO3, IOC_IOMODE_NORMAL);
-			IOCPortConfigureSet(BOARD_RF_SUB1GHZ,
-					IOC_PORT_RFC_GPO0, IOC_IOMODE_NORMAL);
-		} else {
-			/* RF core active --> Sub-1 GHz */
-			IOCPortConfigureSet(BOARD_RF_HIGH_PA,
-					IOC_PORT_GPIO, IOC_IOMODE_NORMAL);
-			IOCPortConfigureSet(BOARD_RF_SUB1GHZ,
-					IOC_PORT_GPIO, IOC_IOMODE_NORMAL);
-			GPIO_setOutputEnableDio(BOARD_RF_SUB1GHZ, GPIO_OUTPUT_ENABLE);
+		if (sub1GHz) {
+			if (paType == RF_TxPowerTable_HighPA) {
+				/* Note: RFC_GPO3 is a work-around because the RFC_GPO1 */
+				/* is sometimes not de-asserted on CC1352 Rev A. */
+				pinctrl_apply_state(ant_pcfg, PINCTRL_STATE_ANT_SUBG_PA);
+			} else {
+				pinctrl_apply_state(ant_pcfg, PINCTRL_STATE_ANT_SUBG);
+				/* Manually set the sub-GHZ antenna switch DIO */
+				gpio_pin_configure_dt(&ant_gpios[BOARD_ANT_GPIO_SUBG], 1);
+			}
```

Comment:
I guess the pinmux must be used because of some silicon error. I'm still confused on that vs. driving low, but it feels very off to use a DT defined pinmux state and then local define statements for GPIO indexes.

Reply:
See https://github.com/zephyrproject-rtos/zephyr/pull/73488#discussion_r1665473681